### PR TITLE
Update workflow versioning

### PR DESCRIPTION
**Pull request by SuperQ**
### 1. Why is this pull request needed and what does it do?
* Use major version pinning for actions to reduce dependeabot churn.
* Use tilde matching for Go versions to use the latest patch release.
* Checkout before installing Go to fix go module caching.

### CODE REVIEW

**Comment by SuperQ** [comment]
Moved the checkout action to first so that caching in setup-go works.

```
Warning: Restore cache failed: Dependencies file is not found in /home/runner/work/coredns/coredns. Supported file pattern: go.sum
```

File: .github/workflows/codeql-analysis.yml [hunk 1]
```diff
@@ -27,7 +27,7 @@ jobs:
 
     steps:
       - name: Checkout repository
-        uses: actions/checkout@93ea575cb5d8a053eaa0ac8fa3b40d7e05a33cc8
+        uses: actions/checkout@v3
 
       - name: Initialize CodeQL
         uses: github/codeql-action/init@cdcdbb579706841c47f7063dda365e292e5cad7a
```
Comment (Tantalor93) [inline .github/workflows/codeql-analysis.yml]: what about these codeql-action versions?

Comment (SuperQ) [reply .github/workflows/codeql-analysis.yml]: Sure

File: .github/workflows/go.coverage.yml [hunk 2]
```diff
@@ -8,15 +8,15 @@ jobs:
     name: Coverage
     runs-on: ubuntu-latest
     steps:
+      - name: Check out code
```
Comment (Tantalor93) [inline .github/workflows/go.coverage.yml:11]: can we also pin `codecov/codecov-action`

File: .github/workflows/scorecards.yml [hunk 3]
```diff
@@ -23,7 +23,7 @@ jobs:
 
     steps:
       - name: "Checkout code"
-        uses: actions/checkout@93ea575cb5d8a053eaa0ac8fa3b40d7e05a33cc8
+        uses: actions/checkout@v3
```
Comment (Tantalor93) [inline .github/workflows/scorecards.yml:26]: can we also pin version actions/upload-artifact and ossf/scoreboards action ?

File: .github/workflows/yamllint.yml [hunk 4]
```diff
@@ -8,7 +8,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - name: 'Checkout'
-        uses: actions/checkout@93ea575cb5d8a053eaa0ac8fa3b40d7e05a33cc8
+        uses: actions/checkout@v3
```
Comment (Tantalor93) [inline .github/workflows/yamllint.yml:11]: can we pin yamllint action also?

**Review by Tantalor93 — COMMENTED** [review]
nice! this looks very useful, i would just make sure we do this consistently for all workflows and their used actions

**Review by bradbeam — APPROVED** [review]

File: .github/workflows/yamllint.yml (hunk 4 above)
Comment (SuperQ) [reply .github/workflows/yamllint.yml:11]: That one doesn't seem to promote this kind of pinning, can try tho.

Comment (SuperQ) [reply .github/workflows/yamllint.yml:11]: Nope, this one didn't work.

**Review by Tantalor93 — APPROVED** [review]

**Comment by SuperQ** [comment]
Now, hopefully dependabot doesn't just revert all these version pins. :grin:

**Pull request merged**